x-pack/elastic-agent/pkg/agent/application/handler_action_application.go

```diff
@@ -23,9 +28,41 @@ func (h *handlerAppAction) Handle(ctx context.Context, a action, acker fleetAcke
 		return fmt.Errorf("invalid type, expected ActionApp and received %T", a)
 	}
 
-	_ = action
+	appState, ok := h.srv.FindByInputType(action.InputType)
+	if !ok {
+		return fmt.Errorf("matching app is not found for action input: %s", action.InputType)
+	}
+
+	params, err := action.MarshalMap()
+	if err != nil {
+		return err
+	}
 
-	// TODO: handle app action
+	start := time.Now().UTC().Format(time.RFC3339Nano)
+	res, err := appState.PerformAction(action.InputType, params, defaultActionTimeout)
+	end := time.Now().UTC().Format(time.RFC3339Nano)
+	if err != nil {
+		action.StartedAt = start
+		action.CompletedAt = end
+		action.Error = err.Error()
+	} else {
+		action.StartedAt = readMapString(res, "started_at")
+		action.CompletedAt = readMapString(res, "completed_at")
+		action.Error = readMapString(res, "error")
```

Comment:
What if these keys are empty?

Reply:
probably ok for the error, can add some value for started/completed